src/js/components/SelectMultiple/SelectMultiple.js

```diff
@@ -0,0 +1,529 @@
+    // what we send with onChange
+    // When 'valueKey' sets 'reduce', the value(s) here should match
+    // what the 'valueKey' would return for the corresponding
+    // selected option object.
+    // Otherwise, the value(s) should match the selected options.
+
+    const [value, setValue] = formContext.useFormInput({
+      name,
+      value: valueProp,
+      initialValue: defaultValue || '',
+    });
+
+    // normalizedValue is the value mapped with any valueKey applied
+    // When the options array contains objects, this property indicates how
+    // to retrieve the value of each option.
+    // If a string is provided, it is used as the key to retrieve a
+    // property of an option object.
+    // If a function is provided, it is called with the option and should
+    // return the value.
+    // If reduce is true, this value will be used for the 'value'
+    // delivered via 'onChange'.
+    const normalizedValue = useMemo(
+      () => getNormalizedValue(value, valueKey),
+      [value, valueKey],
+    );
+    // search input value
+    const [search, setSearch] = useState();
+    // All select option indices and values
+    const [allOptions, setAllOptions] = useState(optionsProp);
+    const [orderedOptions, setOrderedOptions] = useState();
+    // Track changes to options property, except when options are being
+    // updated due to search activity. Allows option's initial index value
+    // to be referenced when filtered by search.
+    useEffect(() => {
+      if (!search) setAllOptions(optionsProp);
+      if (!search && separateSelected) setOrderedOptions(optionsProp);
+    }, [optionsProp, search, separateSelected]);
+
+    // the option indexes present in the value
+    const optionIndexesInValue = useMemo(() => {
+      const result = [];
+      allOptions.forEach((option, index) => {
+        if (
+          normalizedValue &&
+          normalizedValue.length > 0 &&
+          normalizedValue.some((v) => v === applyKey(option, valueKey))
+        ) {
+          result.push(index);
+        }
+      });
+      return result;
+    }, [allOptions, valueKey, normalizedValue]);
+
+    const [open, setOpen] = useState(openProp);
+    useEffect(() => setOpen(openProp), [openProp]);
+
+    const onRequestOpen = useCallback(() => {
+      if (open) return;
+      setOpen(true);
+      if (onOpen) onOpen();
+    }, [onOpen, open]);
+
+    useEffect(() => {
+      if (separateSelected && ((open && search) || !open)) {
+        const selectedOptions = optionsProp.filter((option) => {
+          const iValue =
+            valueKey && valueKey.reduce ? applyKey(option, valueKey) : option;
+          return value.includes(iValue);
+        });
+        const unselectedOptions = optionsProp.filter(
+          (i) => !selectedOptions.includes(i),
+        );
+        const nextOrderedOptions = selectedOptions.concat(unselectedOptions);
+        setOrderedOptions(nextOrderedOptions);
```

Comment:
Should `orderedOptions` just be calculated via `useMemo()` when any dependencies change, as opposed to needing `useState()` + `useEffect()`?

Reply:
I played around with using `useMemo` for this but wasn't getting the results I wanted. `orderedOptions` was getting reset to the regular list of options at the incorrect times so that the ordering wasn't reflecting in the list. I did tweak the useEffect slightly to tighten things up.